localstack/aws/serving/twisted.py

```diff
@@ -0,0 +1,290 @@
+        headerSequence = [responseLine]
+
+        for name, value in headers:
+            line = name + b": " + value + b"\r\n"
+            headerSequence.append(line)
+
+        headerSequence.append(b"\r\n")
+        self.transport.writeSequence(headerSequence)
+
+    def isSecure(self):
+        # used to determine the WSGI url scheme (http vs https)
+        try:
+            # ``self.transport`` will be a ``TLSMultiplexer`` instance in our case
+            return self.transport.isSecure()
+        except AttributeError:
+            return super().isSecure()
+
+
+class TLSMultiplexer(ProtocolWrapper):
+    """
+    Custom protocol to multiplex HTTPS and HTTP connections over the same port. This is the equivalent of
+    ``DuplexSocket``, but since twisted use its own SSL layer and doesn't use `ssl.SSLSocket``, we need to implement
+    the multiplexing behavior in the Twisted layer.
+
+    The basic idea is to defer the ``makeConnection`` call until the first data are received, and then re-configure
+    the underlying ``wrappedProtocol`` if needed with a TLS wrapper.
+    """
+
+    tlsProtocol = BufferingTLSTransport
+
+    def __init__(
+        self,
+        factory: "WrappingFactory",
+        wrappedProtocol: interfaces.IProtocol,
+    ):
+        super().__init__(factory, wrappedProtocol)
+        self._isInitialized = False
+        self._isTLS = None
+        self._negotiatedProtocol = None
+
+    def makeConnection(self, transport):
+        self.connected = 1
+        self.transport = transport
+        # we defer the actual makeConnection call to the first invocation of dataReceived
+
+    def dataReceived(self, data: bytes) -> None:
+        if self._isInitialized:
+            super().dataReceived(data)
+            return
+
+        # once the first data have been received, we can check whether it's a TLS handshake, then we need to run the
+        # actual makeConnection procedure.
+        self._isInitialized = True
+        self._isTLS = data[0] == 22  # 0x16 is the marker byte identifying a TLS handshake
+
+        if self._isTLS:
+            # wrap protocol again in tls protocol
+            self.wrappedProtocol = self.tlsProtocol(self.factory, self.wrappedProtocol)
+        else:
+            if data.startswith(b"PRI * HTTP/2"):
+                # TODO: can we do proper protocol negotiation like in ALPN?
+                # in the TLS case, this is determined by the ALPN procedure by OpenSSL.
+                self._negotiatedProtocol = b"h2"
+
+        # now that we've set the real wrapped protocol, run the make connection procedure
+        super().makeConnection(self.transport)
+        super().dataReceived(data)
+
+    @property
+    def negotiatedProtocol(self) -> str | None:
+        if self._negotiatedProtocol:
+            return self._negotiatedProtocol
+        return self.wrappedProtocol.negotiatedProtocol
+
+
+class TLSMultiplexerFactory(TLSMemoryBIOFactory):
```

Comment:
It seems that there is a non-breaking error being logged in our CI when testing our Pro integration:
https://github.com/localstack/localstack/actions/runs/8179057046/job/22366680648?pr=9834

```
Unhandled Error
Traceback (most recent call last):
  File "/home/runner/work/localstack/localstack/localstack/localstack/utils/threads.py", line 60, in run
    result = self.func(self.params, **kwargs)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/internet/base.py", line 695, in run
    self.mainLoop()
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/internet/base.py", line 708, in mainLoop
    self.doIteration(t)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/internet/epollreactor.py", line 244, in doPoll
    log.callWithLogger(selectable, _drdw, selectable, fd, event)
--- <exception caught here> ---
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/python/log.py", line 96, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/python/log.py", line 80, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/python/context.py", line 117, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/python/context.py", line 82, in callWithContext
    return func(*args, **kw)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/internet/posixbase.py", line 495, in _doReadOrWrite
    self._disconnectSelectable(selectable, why, inRead)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/internet/posixbase.py", line 107, in _disconnectSelectable
    selectable.readConnectionLost(f)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/internet/tcp.py", line 311, in readConnectionLost
    self.connectionLost(reason)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/internet/tcp.py", line 328, in connectionLost
    protocol.connectionLost(reason)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/protocols/policies.py", line 113, in connectionLost
    self.factory.unregisterProtocol(self)
  File "/home/runner/work/localstack/localstack/localstack-ext/.venv/lib/python3.11/site-packages/twisted/protocols/policies.py", line 167, in unregisterProtocol
    del self.protocols[p]
builtins.KeyError: <localstack.aws.serving.twisted.TLSMultiplexer object at 0x7f90ef16ea10>
```

Reply:
good catch! found the issue and it's fixed